### ISSUE TITLE
[Bug]: Vite builder assumes that root directory is parent directory of config dir

### ISSUE THREAD

**Issue by stubhubstevezhu** [storybookjs/storybook#22508]
### Describe the bug

Vite builder assumes that root directory for vite is the parent directory of the config directory:
https://github.com/storybookjs/storybook/blob/v7.0.10/code/lib/builder-vite/src/vite-config.ts#L55

However, you can have a directory layout like `.storybook/custom/main.ts` with the command `storybook dev -c .storybook/custom` which would cause the root directory to be set improperly.

### System

```shell
Environment Info:

  System:
    OS: Windows 10 10.0.19044
    CPU: (8) x64 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
  Binaries:
    Node: 19.7.0 - C:\Program Files\nodejs\node.EXE
    npm: 9.6.5 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (113.0.1774.35)
  npmPackages:
    @storybook/addon-essentials: ^7.0.8 => 7.0.8
    @storybook/addon-interactions: ^7.0.8 => 7.0.8
    @storybook/addon-links: ^7.0.8 => 7.0.8
    @storybook/addon-storysource: ^7.0.8 => 7.0.8
    @storybook/react: ^7.0.8 => 7.0.8
    @storybook/react-vite: ^7.0.8 => 7.0.8
    @storybook/testing-library: ^0.1.0 => 0.1.0
```

**Comment by IanVS** [comment]
@stzhush thanks for the report.  You're correct that we make this assumption, which usually tends to work out correctly.  Can you talk a bit about your use case for having a custom storybook config directory?  I'd like to make sure that whatever solution we establish will work for all users.

Note that in the meantime, you can set the `root` in `viteFinal` in your storybook config, so that it's correct.

**Comment by stzhush** [comment]
@IanVS I was playing around with composing local storybooks following the following guide to split the storybook into a storybook for the design system and a storybook for application specific components.
https://storybook.js.org/docs/react/sharing/storybook-composition#compose-local-storybooks.

But in general if we have the config option `--config-dir` to configure the storybook configuration directory, it's not obvious that it can only be at a certain nesting level. I think it's fine if the assumption that the root directory for vite is the parent directory of the config directory is kept, but it might be helpful to output an error if anyone attempts to set it to something else. Otherwise it might be good to figure out the `root` option through another method.

In the case where we can't nest the storybook configs, I guess I would probably setup the project with something like `projectRoot/.storybook-shared/` and `projectRoot/.storybook-app/`.

**Comment by IanVS** [comment]
One other idea is to use [find-up](https://github.com/sindresorhus/find-up) to search for the directory containing either a `package.json` or `vite.config` file.  Or alternatively, we could use cwd, like Vite itself does.  But, I worry that in each case, there will be users for whom this does not yield the correct result.  The _vast_ majority of storybook projects use `.storybook` as the config directory, so this heuristic of "the parent of .storybook" has worked out well for a couple of years now.  I'm a bit nervous to change it, to be honest.  I wonder if the best thing is to just more thoroughly document that "if you change the configDir, and are using a vite framework, also set the `root` in `viteFinal`.  🤔

**Comment by stzhush** [comment]
Yeah I don't necessarily prefer nesting my storybook configs. I just submitted a bug report because what I assumed would be a valid configuration from the documentation didn't work for unexpected reasons. It seems like if a fix on the implementation side seems a bit iffy, it could be good to improve the documentation. But I don't really think the suggestion is as easy as "if you change the configDir, and are using a vite framework, also set the root in viteFinal". There probably are (or will be) various other assumptions in the codebase about the location of the storybook config directory other than just the `root` config in vite. It seems better to just suggest that projects keep the config in the root directory if that's the assumption being held for implementation.

Though one question I'm also wondering about is what is the intended use case of `--config-dir`?

**Comment by IanVS** [comment]
In terms of the Vite builder, setting of the `root` is the only spot where we kind of ignored the config dir setting.  I'm pretty sure the rest of Storybook works correctly with it set.  

As to the purpose of that setting, it's a good question.  Presumably it's used by folks with some some kind of esoteric monorepo setup.  The topic of best practices for Storybook in monorepos came up earlier today as well, and it's something I hope we can nail down.  @shilman do we have any telemetry on how folks are using the config dir flag, so we can try to make sure to support them in this way?

**Comment by Rulexec** [comment]
It also breaks aliases like this:

```
resolve: {
  alias: {
    '@': '/src',
  },
},
```

I replaced it with `'@': resolve(__dirname, 'src'),` and all was fixed, but it looks inconvinient that I cannot easily put storybook configs under `.config/storybook` instead of just `.storybook` to not flood root dir too much.
